### ISSUE TITLE
RESTDataSource is not putting response in redis cache

### ISSUE THREAD

**Issue by uarsovic** [apollographql/apollo-server#1460]
RESTDataSource is not putting response in redis cache when the response header has Cache-control set to private. It is storing it in internal key/value map.

The bug itself I think is in this part of code i HTTPCache class:
"if (!ttlOverride && !(request.method === 'GET' && policy.storable()))
      return response;"
Line number 100.

**Comment by martijnwalraven** [comment]
That is expected behavior. The cache is shared between users, so private data shouldn't be stored there.

But if the response sets `Cache-Control` to `private` by mistake, you can override it by setting an explicit `ttl` (`this.get('/post', id, { ttl: 3600 })`.

**Comment by uarsovic** [comment]
@martijnwalraven Thx, yeah it is working now with slight changes e.g. ".get(requestUrl, null, {   
 cacheOptions: {ttl: 300}})"
Another question is, why do you use same value for the Redis expiration? It will for instance override default value of the Redis server?

**Comment by martijnwalraven** [comment]
> Another question is, why do you use same value for the Redis expiration? It will for instance override default value of the Redis server?

Could you elaborate on what you mean by this? We use the `ttl` value to set the time to live of the cache entry. Is that not what you would expect?

**Comment by uarsovic** [comment]
In this line of code: "await this.keyValueCache.set(`httpcache:${cacheKey}`, entry, {ttl,})" we are setting same ttl value for redis and for the get request. It would be nice if we have some true/false parameter to say not to use it for redis also.

**Comment by martijnwalraven** [comment]
I'm not sure I understand what you mean. The `ttl` value you set is meant to specify how long you want the request you're performing to be cached. So it makes sense to me that would also define the `ttl` for Redis. What is your use case for wanting to separate the two?

(Note that we don't use the set TTL directly, but by default cache entries twice as long, to allow for conditional requests with ETags to revalidate stale responses.)

**Comment by uarsovic** [comment]
So, our use case is. We do a get request to 3rd party and we want to put the response to redis cache/DB. But we don't want to have ttl for redis entry. We want that value to be in redis until we decide to flush redis entry or whole cache/DB. If we don't pass the ttl argument to get method it won't be put in redis.

**Comment by martijnwalraven** [comment]
Ah, note that just not setting the `ttl` in Redis wouldn't help here. `RESTDataSource` will still validate the max age of the cached response against the `ttl` you specified, and won't return it when it is stale.

It seems what you're looking for is a way to specify a `ttl` value that means 'never expire'. Maybe something like `ttl: 'indefinite'`?

**Comment by uarsovic** [comment]
Yeah, something like that, or if we don't specify ttl it will be indefinite. Since redis default value for cache is indefinite.

**Comment by martijnwalraven** [comment]
But that won't work, because not specifying a TTL override means respecting the HTTP caching headers. And in your case, the response returning `Cache-Control: private` would turn off caching altogether.

**Comment by ajkotsubailo** [comment]
@martijnwalraven Thank you

**Comment by zjovan** [comment]
@martijnwalraven 

> That is expected behavior. The cache is shared between users, so private data shouldn't be stored there. But if the response sets Cache-Control to private by mistake, you can override it by setting an explicit ttl, e.g. this.get('/post', id, { ttl: 3600 }).

If ttl is overridden, then all responses will be cached e.g. policy.storable() returns non-storable cache status code 404, 403, 500 that response will be stored?

**Comment by martijnwalraven** [comment]
Right now that is the behavior I believe, but it sounds like something we may want to change.

**Comment by ericclemmons** [comment]
This is an interesting thread. I've noticed that the default `InMemoryLRUCache` and `RedisCache` both haven't been returning responses from `this.httpCache`:

1. `this.get` <https://api.github.com/repos/apollographql/apollo-server> & notice `Cache-control: public, max-age=60, s-maxage=60`.
2. Confirm rate limit: <https://api.github.com/rate_limit>
3. `this.httpCache` (when using `InMemoryLRUCache` correctly sets a `httpcache:...` key/value pair).
4. Subsequent `this.get` requests make `fetch` request, as evidenced by <https://api.github.com/rate_limit>.

**Comment by martijnwalraven** [comment]
@ericclemmons Maybe I misunderstand your issue, but it seems https://api.github.com/rate_limit returns `Cache-Control: no-cache`, so that explains why results for that URL aren't cached.

**Comment by ericclemmons** [comment]
Sorry, to be clear I use that URL to confirm the other calls aren't being
cached and impacting my rate limit.

My expectation is that repeated calls would come from the cache and the
rate would decrement only once.

**Comment by martijnwalraven** [comment]
@ericclemmons Ah, I wound indeed expect  https://api.github.com/repos/apollographql/apollo-server to cache for 60 seconds. Is that not the behavior you're seeing? If you run with `NODE_ENV=development`, it should be evident from the timings in the request logs whether a response has been cached.

**Comment by martijnwalraven** [comment]
@ericclemmons What does your data source code look like?

**Comment by ericclemmons** [comment]
I'll put together a small demo, now that I can produce it!

**Comment by martijnwalraven** [comment]
@ericclemmons Any update on this, or can the issue be closed?

**Comment by ericclemmons** [comment]
I still need to split out the code into a demo.

**Comment by abernix** [comment]
Closing this as I feel the original issue is resolved but, @ericclemmons, if you're still experiencing a problem, please put together a reproduction and open a new issue!  Thanks!
